### PR TITLE
initial circuit size estimation added

### PR DESCRIPTION
**Pull request by CblPOK-git**

### CODE REVIEW

File: include/nil/blueprint/handle_component.hpp [hunk 1]
```diff
@@ -61,11 +61,18 @@
 #include <nil/blueprint/asserts.hpp>
 #include <nil/blueprint/stack.hpp>
 #include <nil/blueprint/policy/policy_manager.hpp>
+#include <typeinfo>
 
 
 namespace nil {
     namespace blueprint {
 
+        std::map<std::string, int> components_frequency;
```
Comment (akokoshn) [inline include/nil/blueprint/handle_component.hpp:70]: 1. I suggest use map<sting, statistic>
```
struct statistic {
int components_frequency;
int rows_used;
int gates_used
}
```
2. Global variable may be not better design solution. I guggest to keep map in parser instance and pass as reference.

File: include/nil/blueprint/handle_component.hpp [hunk 2]
```diff
@@ -61,11 +61,18 @@
...
+        std::map<std::string, int> rows_used;
+        std::map<std::string, int> rows_per_component;
```
Comment (akokoshn) [inline include/nil/blueprint/handle_component.hpp:72]: what is different `rows_used` vs `rows_per_component`?

File: include/nil/blueprint/statistics.hpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                "non_native fp12 multiplication",
+                "is_in_g1",
+                "is_in_g2",
+                "native_bls12_381_pairing",
+                "hash to curve",
+                "comparison",
+            };
+
+            void add_record(std::string name, std::size_t rows, std::size_t gates, std::size_t witness) {
+                component_counter[name] += 1;
```
Comment (akokoshn) [inline include/nil/blueprint/statistics.hpp:50]: `if (find(name) != end) {set counter  = 1, rows, withness, ...,} else {update counter}`

File: include/nil/blueprint/statistics.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                component_witness[name] = witness;
+                if (unfinished_components.find(name) != unfinished_components.end()) {
+                    component_finished[name] = false;
+                } else {
+                    component_finished[name] = true;
+                }
+            }
+
+            void print() {
+                std::cout << "\n=====================================================================================\n";
```
Comment (akokoshn) [inline include/nil/blueprint/statistics.hpp:62]: looks not so good in small console (takes few lines), may be short delimiter enough `\b====\n`

**Review by akokoshn — CHANGES_REQUESTED** [review]

**Review by makxenov — APPROVED** [review]

File: include/nil/blueprint/statistics.hpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                    std::size_t total_rows = component.component_rows * component.component_counter;
+                    std::cout << " (" << total_rows << " in total)\n";
+                    std::cout << "________________\n";
+                }
+                std::cout << std::endl;
+
+            }
+        };
+
+        nil::blueprint::component_calls statistics;
```
Comment (akokoshn) [inline include/nil/blueprint/statistics.hpp:116]: global variable is not so good solution, can we keep `statistics` as member of `parser` and pass by reference like `assignment table`?

**Review by akokoshn — CHANGES_REQUESTED** [review]

**Review by akokoshn — APPROVED** [review]

**Pull request merged**